### PR TITLE
Force consistent delay for stream playback

### PR DESCRIPTION
**Pull request by StephenCWills**
The idea here is to force the delay for stream playback to be equal to the length of the audio buffer. This is achieved by applying an offset to the first frame of playback based on the amount of time elapsed between the start of playback and the subsequent audio frame callback. This provides a more consistent delay and smooths out the unevenness in playback that comes about from always aligning the start of each stream to the start of the next audio buffer.

Testing reveals this fixes the uneven playback delays discovered in https://github.com/diasurgical/devilutionX/issues/1852. That said, I have little confidence in my understanding of SDL audio so I think it would be wise to give this a thorough review.

### CODE REVIEW

File: src/stream_p.cpp [hunk 1]
```diff
@@ -111,6 +111,9 @@ void Aulib::Stream_priv::fSdlCallbackImpl(void* /*unused*/, Uint8 out[], int out
         return fStreamList;
     }();
 
+    int now = SDL_GetTicks();
+    int bufferMilliseconds = fAudioSpec.samples * 1000 / fAudioSpec.freq;
```
Comment (realnc) [inline src/stream_p.cpp]: Is it guaranteed that SDL will not pass a buffer that's shorter/longer than `fAudioSpec.samples`? If not, then this should be used instead:

```
int bufferMilliseconds = (wantedSamples / fAudioSpec.channels) * 1000 / fAudioSpec.freq;
```

File: src/stream_p.cpp [hunk 2]
```diff
@@ -120,25 +123,39 @@ void Aulib::Stream_priv::fSdlCallbackImpl(void* /*unused*/, Uint8 out[], int out
...
+
         bool has_finished = false;
         bool has_looped = false;
-        int len = 0;
+        int offset = 0;
+        int end = 0;
+
+        if (diffMilliseconds < bufferMilliseconds) {
+            int offsetMilliseconds = bufferMilliseconds - diffMilliseconds;
+            offset = wantedSamples * offsetMilliseconds / bufferMilliseconds;
```
Comment (realnc) [inline src/stream_p.cpp]: I believe this should be:

```
offset = offsetMilliseconds * fAudioSpec.channels * fAudioSpec.freq / 1000;
```

Using the ratio of the two millisecond values seems way too coarse and the resulting offset would end up being quite inaccurate.

File: src/stream_p.cpp [hunk 3]
```diff
@@ -120,25 +123,39 @@ void Aulib::Stream_priv::fSdlCallbackImpl(void* /*unused*/, Uint8 out[], int out
...
+            offset -= offset % 2;
```
Comment (realnc) [inline src/stream_p.cpp]: Is this supposed to round down to fit the channel count? If so, this should be:

```
offset -= offset % fAudioSpec.channels;
```

Comment (StephenCWills) [reply src/stream_p.cpp]: Actually, I had encountered errors on some of the sound effects in DevilutionX in the following section of code.
https://github.com/realnc/SDL_audiolib/blob/7979aa98e3845095c1386b4f1c01e59de3d7d3d3/src/stream_p.cpp#L192-L195

I found that even if SDL was only using one channel, it was still possible for left/right volume to be applied to every other channel.  After that, I got a bit paranoid and started thinking that maybe the number of output channels had nothing to do with the size of the mix buffer. So I simply aligned offset to an even index and waited to see what you'd say about it.

Actually, it looks like the left/right volume is based on position, specifically for stereo sound. Does it make sense to apply left/right volume when channel count is 1? If not, then I think you're right that we should be fitting the channel count, and we should be checking the number of channels before applying left/right volume.

File: src/stream_p.cpp (hunk 1 above)
Comment (StephenCWills) [reply src/stream_p.cpp]: I'm not sure whether it's guaranteed, but my first instinct would be to use your code regardless. I was just paranoid about left/right volume so I made it work with `fAudioSpec.samples` instead.

**Comment by StephenCWills** [comment]
@realnc I just pushed a new commit that applies the three suggestions you provided. To fix the left/right volume errors, I ignored the stereo position if there's only one channel. Although, perhaps it would be better to apply the stereo position modifier to make the sound seem more distant. Let me know what you think.

**Comment by realnc** [comment]
> To fix the left/right volume errors, I ignored the stereo position if there's only one channel. Although, perhaps it would be better to apply the stereo position modifier to make the sound seem more distant. Let me know what you think.

The library does not model sound distance at all, so this is fine.

I pushed a minor fixup and merged the PR. It's just renaming things fixes const correctness (I really need to do this over the whole code base at some point, it's quite inconsistent right now.)

In a followup commit I'll do shortly, I'll fix the case where this fails:

```
if (ticks_since_play_start < wanted_ticks)
```

SDL will call the audio callback during the same millisecond at startup with some audio drivers. PulseAudio for example does this on my Linux box. As a result, there is severe audio crackling for a fraction of a second on streams that start playing immediately after `Aulib::init()`.

It will also apply this delay compensation mechanism to things like `pause()` and `rewind()` as well.

**Comment by StephenCWills** [comment]
Actually, const correctness is one of my weak points as I don't use C++ very often. I certainly appreciate all your help, and I'm glad we were able to find a fix for this.

**Pull request merged**